ray-operator/test/e2e/rayservice_ha_test.go

```diff
@@ -200,3 +201,73 @@ func TestRayServiceZeroDowntimeUpgrade(t *testing.T) {
 		"python", "/locust-runner/locust_runner.py", "-f", "/locustfile/locustfile.py", "--host", "http://test-rayservice-serve-svc:8000",
 	})
 }
+
+func TestRayServiceGCSFaultTolerance(t *testing.T) {
+	rayserviceYamlFile := "testdata/rayservice.ft.yaml"
+	locustYamlFile := "testdata/locust-cluster.const-rate.yaml"
+
+	test := With(t)
+	g := NewWithT(t)
+
+	// Create a namespace
+	namespace := test.NewTestNamespace()
+	test.StreamKubeRayOperatorLogs()
+
+	// Create a ConfigMap with Locust runner script
+	configMapAC := newConfigMap(namespace.Name, "locust-runner-script", files(test, "locust_runner.py"))
+	configMap, err := test.Client().Core().CoreV1().ConfigMaps(namespace.Name).Apply(test.Ctx(), configMapAC, TestApplyOptions)
+	g.Expect(err).NotTo(HaveOccurred())
+	test.T().Logf("Created ConfigMap %s/%s successfully", configMap.Namespace, configMap.Name)
+
+	// Create the RayService for testing
+	KubectlApplyYAML(test, rayserviceYamlFile, namespace.Name)
+	rayService, err := GetRayService(test, namespace.Name, "test-rayservice")
+	g.Expect(err).NotTo(HaveOccurred())
+	test.T().Logf("Created RayService %s/%s successfully", rayService.Namespace, rayService.Name)
+
+	test.T().Logf("Waiting for RayService %s/%s to running", rayService.Namespace, rayService.Name)
+	g.Eventually(RayService(test, rayService.Namespace, rayService.Name), TestTimeoutMedium).
+		Should(WithTransform(RayServiceStatus, Equal(rayv1.Running)))
+
+	// Get the underlying RayCluster of the RayService
+	rayService, err = GetRayService(test, namespace.Name, rayService.Name)
+	g.Expect(err).NotTo(HaveOccurred())
+	rayServiceUnderlyingRayCluster, err := GetRayCluster(test, namespace.Name, rayService.Status.ActiveServiceStatus.RayClusterName)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// Create Locust RayCluster
+	KubectlApplyYAML(test, locustYamlFile, namespace.Name)
+	locustCluster, err := GetRayCluster(test, namespace.Name, "locust-cluster")
+	g.Expect(err).NotTo(HaveOccurred())
+	test.T().Logf("Created Locust RayCluster %s/%s successfully", locustCluster.Namespace, locustCluster.Name)
+
+	g.Eventually(RayCluster(test, locustCluster.Namespace, locustCluster.Name), TestTimeoutMedium).
+		Should(WithTransform(RayClusterState, Equal(rayv1.Ready)))
+	g.Expect(GetRayCluster(test, locustCluster.Namespace, locustCluster.Name)).To(WithTransform(RayClusterDesiredWorkerReplicas, Equal(int32(0))))
+
+	locustHeadPod, err := GetHeadPod(test, locustCluster)
+	g.Expect(err).NotTo(HaveOccurred())
+	test.T().Logf("Found head pod %s/%s", locustHeadPod.Namespace, locustHeadPod.Name)
+
+	// Install Locust in the Locust head Pod
+	ExecPodCmd(test, locustHeadPod, common.RayHeadContainer, []string{"pip", "install", "locust"})
+
+	// Get no ops' head pod
+	noOpsHeadPod, err := GetHeadPod(test, rayServiceUnderlyingRayCluster)
+	g.Expect(err).NotTo(HaveOccurred())
+	// Kill gcs server
+	ExecPodCmd(test, noOpsHeadPod, common.RayHeadContainer, []string{"pkill", "gcs_server"})
```

Comment:
Maybe we can change the implementation to below, so we don't need a goroutine.

* Store the name of the head Pod in a variable (e.g., `oldHeadPodName`).
* Kill the GCS process on the old head Pod.
* Use `ExecPodCmd` to submit requests.
* After ExecPodCmd finishes, check the following:
  * Whether `ExecPodCmd` ran for more than 2 minutes. Because this test shares the Locust RayCluster YAML file with other tests, we need to ensure the YAML file is not accidentally updated.
  * Check the name of the head Pod which should remain the same as `oldHeadPodName`.
  * The number of `Restarts` for the head Pod should be 1.
  * All Pods are running.

Reply:
OK, no problem